lib/reversetunnel/remoteclientmanager.go

```diff
@@ -0,0 +1,308 @@
+	}
+	return m, nil
+}
+
+// connect runs until all clients have been successfully setup.
+func (m *remoteClientManager) connect(ctx context.Context) error {
+	var (
+		client      auth.ClientI
+		accessPoint auth.RemoteProxyAccessPoint
+		nodeWatcher *services.NodeWatcher
+		caWatcher   *services.CertAuthorityWatcher
+		err         error
+	)
+
+	firstIteration := true
+	for {
+		if err := ctx.Err(); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Use backoff after first attempt.
+		if !firstIteration {
+			<-m.backoff.After()
+			m.backoff.Inc()
+		} else {
+			firstIteration = false
+		}
+
+		if client == nil {
+			client, err = m.newClientFunc(ctx)
+			if err != nil {
+				m.log.WithError(err).Warnf("Failed to connect to remote auth server.")
+				continue
+			}
+		}
+
+		response, err := client.Ping(ctx)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to get remote auth server version.")
+			continue
+		}
+
+		version := response.ServerVersion
+		if accessPoint != nil {
+			accessPoint.Close()
+		}
+		accessPoint, err = m.newAccessPointFunc(ctx, client, version)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote access point.")
+			continue
+		}
+
+		if nodeWatcher != nil {
+			nodeWatcher.Close()
+		}
+		nodeWatcher, err = m.newNodeWatcherFunc(ctx, accessPoint)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote node watcher.")
+			continue
+		}
+
+		if caWatcher != nil {
+			caWatcher.Close()
+		}
+		caWatcher, err = m.newCAWatcher(ctx, accessPoint)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote CA watcher.")
+			continue
+		}
+
+		m.clientLock.Lock()
+		m.auth = client
+		m.accessPoint = accessPoint
+		m.nodeWatcher = nodeWatcher
+		m.caWatcher = caWatcher
+		m.once.Do(func() {
+			close(m.connectedOnceOrClosed)
+		})
+		m.clientLock.Unlock()
+
+		return nil
+	}
+}
+
+// Connect blocks while all clients are created. If connect is called more than once
+// existing clients will be closed and new clients created. Connect will always fail
+// after Close is called.
+func (m *remoteClientManager) Connect() error {
+	errC := make(chan error)
+	m.connectLock.Lock()
+	if m.closing {
+		m.connectLock.Unlock()
+		return trace.Errorf("unable to connect: auth manager is closing")
+	}
+
+	// Stop the current connection attempt.
+	if m.stop != nil {
+		m.stop()
+		m.wg.Wait()
+	}
+
+	ctx, cancel := context.WithCancel(m.parentCtx)
+	m.stop = cancel
+
+	m.clientLock.Lock()
+	client := m.auth
+	accessPoint := m.accessPoint
+	nodeWatcher := m.nodeWatcher
+	caWatcher := m.caWatcher
+	m.auth = nil
+	m.accessPoint = nil
+	m.nodeWatcher = nil
+	m.caWatcher = nil
+
+	// Close previous clients in background.
+	go func() {
+		m.close(client, accessPoint, nodeWatcher, caWatcher)
+	}()
+	m.clientLock.Unlock()
+
+	m.wg.Add(1)
+	m.connectLock.Unlock()
+	go func() {
+		defer m.wg.Done()
+		errC <- m.connect(ctx)
+	}()
+	return trace.Wrap(<-errC)
```

Comment:
This is a very strange way of doing `defer m.wg.Done(); return trace.Wrap(m.connect(ctx))`. Why do we need to bump and reduce the waitgroup?